data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,221 @@
+    private final String logGroup;
+    private final String logStream;
+    private final int retryCount;
+    private final long backOffTimeBase;
+    private final io.micrometer.core.instrument.Counter logEventSuccessCounter; //Counter to be used on the fly for counting successful transmissions. (Success per single event successfully published).
+    private final Counter requestSuccessCount;
+    private final io.micrometer.core.instrument.Counter logEventFailCounter;
+    private final io.micrometer.core.instrument.Counter requestFailCount; //Counter to be used on the fly during error handling.
+    private int failCounter = 0;
+    private boolean failedPost;
+    private final StopWatch stopWatch;
+    private boolean stopWatchOn;
+    private final ReentrantLock reentrantLock;
+
+    public CloudWatchLogsService(final CloudWatchLogsClient cloudWatchLogsClient, final CloudWatchLogsSinkConfig cloudWatchLogsSinkConfig, final Buffer buffer,
+                                 final PluginMetrics pluginMetrics, final ThresholdCheck thresholdCheck, final int retryCount, final long backOffTimeBase) {
+
+        this.cloudWatchLogsClient = cloudWatchLogsClient;
+        this.buffer = buffer;
+        this.logGroup = cloudWatchLogsSinkConfig.getLogGroup();
+        this.logStream = cloudWatchLogsSinkConfig.getLogStream();
+        this.thresholdCheck = thresholdCheck;
+
+        this.retryCount = retryCount;
+        this.backOffTimeBase = backOffTimeBase;
+
+        this.bufferedEventHandles = new ArrayList<>();
+        this.logEventSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_SUCCESS);
+        this.requestFailCount = pluginMetrics.counter(REQUESTS_FAILED);
+        this.logEventFailCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_FAIL);
+        this.requestSuccessCount = pluginMetrics.counter(REQUESTS_SUCCEEDED);
+
+        reentrantLock = new ReentrantLock();
+
+        stopWatch = StopWatch.create();
+        stopWatchOn = false;
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void output(final Collection<Record<Event>> logs) {
+        reentrantLock.lock();
+
+        try {
+            if (!stopWatchOn) {
+                startStopWatch();
+            }
+
+            for (Record<Event> singleLog: logs) {
+                String logJsonString = singleLog.getData().toJsonString();
+                int logLength = logJsonString.length();
+
+                if (thresholdCheck.isGreaterThanMaxEventSize(logLength + LOG_EVENT_OVERHEAD_SIZE)) {
+                    LOG.warn("Event blocked due to Max Size restriction! {Event Size: " + (logLength + LOG_EVENT_OVERHEAD_SIZE) + " bytes}");
```

Comment:
The most likely scenario is during the DLQ implementation, we can consider pushing the log event to it instead, otherwise the log event cannot meet restrictions and needs to be dropped. A warning metric could be added here as well to track the number of oversizedlogs.